File: programmers/level2/뉴스클러스터링/java/20221230/src/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public int solution(String str1, String str2) {
        int answer;

        List<String> string1 = splitWords(str1);
        List<String> string2 = splitWords(str2);

        List<String> intersection = new ArrayList<>();
        List<String> union = new ArrayList<>();

        for (String value : string1) {
            if (string2.remove(value)) {
                intersection.add(value);
            }

            union.add(value);
        }

        union.addAll(string2);

        if (union.isEmpty()) {
            return 65536;
        }

        double jaccardIndex = jaccardSimilarity(intersection.size(), union.size());

        double temp = jaccardIndex * 65536;

        answer = (int) temp;

        return answer;
    }

    public List<String> splitWords(String string) {
        String word = string.toLowerCase();
        List<String> words = new ArrayList<>();

        for (int i = 0; i < word.length() - 1; i += 1) {
            char a = word.charAt(i);
            char b = word.charAt(i + 1);

            if((a >= 'a' && a <= 'z') && (b >= 'a' && b <= 'z')) {
                words.add(String.valueOf(a) + String.valueOf(b));
            }
        }

        return words;
    }

    public double jaccardSimilarity(double a, double b) {
        return a / b;
    }
}

File: programmers/level2/뉴스클러스터링/javascript/20221230/solution.js
function test() {
  console.log(`splitWords('FRANCE'): fr,ra,an,nc,ce = ${splitWords('FRANCE')}`);
  console.log(`solution('FRANCE', 'FRENCH'): 16384 = ${solution('FRANCE', 'FRENCH')}`);
  console.log(`solution('aa1+aa2', 'AAAA12'): 43690 = ${solution('aa1+aa2', 'AAAA12')}`);
  console.log(`solution('handshake', 'shake hands'): 65536 = ${solution('handshake', 'shake hands')}`);
}

test();

function solution(str1, str2) {
  let answer = 0;

  const string1 = splitWords(str1);
  const string2 = splitWords(str2);

  const intersection = [];
  const union = [];

  for (let i = 0; i < string1.length; i += 1) {
    if (string2.indexOf(string1[i]) >= 0) {
      intersection.push(string2.splice(string2.indexOf(string1[i]), 1));
    }

    union.push(string1[i]);
  }

  for (let i = 0; i < string2.length; i += 1) {
    union.push(string2[i]);
  }

  answer = parseInt(65536 * (intersection.length / union.length));
  
  if(isNaN(answer)) {
    return 65536;
  }

  return answer;
}

function splitWords(string) {
  const word = string.toLowerCase();
  const words = [];

  for (let i = 0; i < word.length - 1; i += 1) {
    const a = word.charAt(i);
    const b = word.charAt(i + 1);

    if ((a >= 'a' && a <= 'z') && (b >= 'a' && b <= 'z')) {
      words.push(a + b);
    }
  }

  return words;
}
